Build fails while running bindgen

Just pulled the toolchain and example (using the instructions here: https://github.com/espressif/rust-esp32-example/blob/main/docs/rust-on-xtensa-installation-x86_64-unknown-linux-gnu.md).  When I build for esp32, I get the following:

```
$ idf.py set-target esp32
[output elided...]
$ RUST_BACKTRACE=1 idf.py build
Executing action: all (aliases: build)
Running ninja in directory /home/brian/src/esp32-rust/rust-esp32-example/build
Executing "ninja all"...
[1/14] Performing build step for 'bootloader'
ninja: no work to do.
[2/12] Performing build step for 'rustlib_project'
FAILED: esp-idf/rustlib/stamp/rustlib_project-build esp-idf/rustlib/target/RustApi.h esp-idf/rustlib/target/xtensa-esp32-espidf/release/librustlib.a 
cd /home/brian/src/esp32-rust/rust-esp32-example/components/rustlib && /usr/bin/cmake -E env CARGO_BUILD_TARGET=xtensa-esp32-espidf CARGO_BUILD_TARGET_DIR=/home/brian/src/esp32-rust/rust-esp32-example/build/esp-idf/rustlib/target cargo build --release --features=std
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
   Compiling rustlib v0.1.0 (/home/brian/src/esp32-rust/rust-esp32-example/components/rustlib)
error: failed to run custom build command for `rustlib v0.1.0 (/home/brian/src/esp32-rust/rust-esp32-example/components/rustlib)`

Caused by:
  process didn't exit successfully: `/home/brian/src/esp32-rust/rust-esp32-example/build/esp-idf/rustlib/target/release/build/rustlib-435225cb26b45082/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=/home/brian/src/esp32-rust/rust-esp32-example/build/esp-idf/rustlib/target/RustApi.h

  --- stderr
  thread 'main' panicked at 'libclang error; possible causes include:
  - Invalid flag syntax
  - Unrecognized flags
  - Invalid flag arguments
  - File I/O errors
  - Host vs. target architecture mismatch
  If you encounter an error missing from this list, please file an issue or a PR!', /home/brian/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/ir/context.rs:535:15
  stack backtrace:
     0: rust_begin_unwind
     1: core::panicking::panic_fmt
     2: core::option::expect_failed
     3: core::option::Option<T>::expect
               at /home/runner/work/idf-rust-build/idf-rust-build/library/core/src/option.rs:351:21
     4: bindgen::ir::context::BindgenContext::new
               at /home/brian/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/ir/context.rs:529:13
     5: bindgen::Bindings::generate
               at /home/brian/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/lib.rs:2272:27
     6: bindgen::Builder::generate
               at /home/brian/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/lib.rs:1434:9
     7: build_script_build::run_bindgen
               at ./build.rs:50:20
     8: build_script_build::main
               at ./build.rs:12:5
     9: core::ops::function::FnOnce::call_once
               at /home/runner/work/idf-rust-build/idf-rust-build/library/core/src/ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
ninja: build stopped: subcommand failed.
ninja failed with exit code 1
```

Some other things:
```
$ bindgen --version
bindgen 0.59.1
$ rustc --version
rustc 1.54.0-dev
$ rustc --print target-list | grep esp
riscv32imac-esp-espidf
riscv32imc-esp-espidf
xtensa-esp32-espidf
xtensa-esp32-none-elf
xtensa-esp32s2-espidf
xtensa-esp32s2-none-elf
xtensa-esp8266-none-elf
$ clang --version
clang version 11.0.0 (https://gitlab-ci-token:K3NnnrM4_yiiFxVHhSYU@gitlab.espressif.cn:6688/idf/llvm-project.git 36dbc8b5ee7dae4a4c5bff73efff0ea7ccb3707e)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/xtensa/xtensa-esp32-elf-clang/bin
$ clang -print-targets
  Registered Targets:
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64
    xtensa - Xtensa 32
$ xtensa-esp32-elf-gcc --version
xtensa-esp32-elf-gcc (crosstool-NG esp-2021r1) 8.4.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Also: is that your GitLab auth token in the clang version string?  Might want to rotate those credentials if so...

Same here, macOS Big Sur 11.5.2
```
alex@my-mac rust-esp32-example-main % RUST_BACKTRACE=1 idf.py build
Executing action: all (aliases: build)
Running ninja in directory /Users/alex/Desktop/tmp/rust-esp32-example-main/build
Executing "ninja all"...
[1/14] Performing build step for 'bootloader'
ninja: no work to do.
[2/12] Performing build step for 'rustlib_project'
FAILED: esp-idf/rustlib/stamp/rustlib_project-build esp-idf/rustlib/target/RustApi.h esp-idf/rustlib/target/xtensa-esp32-espidf/release/librustlib.a /Users/alex/Desktop/tmp/rust-esp32-example-main/build/esp-idf/rustlib/stamp/rustlib_project-build /Users/alex/Desktop/tmp/rust-esp32-example-main/build/esp-idf/rustlib/target/RustApi.h /Users/alex/Desktop/tmp/rust-esp32-example-main/build/esp-idf/rustlib/target/xtensa-esp32-espidf/release/librustlib.a 
cd /Users/alex/Desktop/tmp/rust-esp32-example-main/components/rustlib && /usr/local/Cellar/cmake/3.21.2/bin/cmake -E env CARGO_BUILD_TARGET=xtensa-esp32-espidf CARGO_BUILD_TARGET_DIR=/Users/alex/Desktop/tmp/rust-esp32-example-main/build/esp-idf/rustlib/target cargo build --release --features=std
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
'+memctl' is not a recognized feature for this target (ignoring feature)
   Compiling rustlib v0.1.0 (/Users/alex/Desktop/tmp/rust-esp32-example-main/components/rustlib)
error: failed to run custom build command for `rustlib v0.1.0 (/Users/alex/Desktop/tmp/rust-esp32-example-main/components/rustlib)`

Caused by:
  process didn't exit successfully: `/Users/alex/Desktop/tmp/rust-esp32-example-main/build/esp-idf/rustlib/target/release/build/rustlib-5a6f3d9afd6a881d/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=/Users/alex/Desktop/tmp/rust-esp32-example-main/build/esp-idf/rustlib/target/RustApi.h

  --- stderr
  thread 'main' panicked at 'libclang error; possible causes include:
  - Invalid flag syntax
  - Unrecognized flags
  - Invalid flag arguments
  - File I/O errors
  - Host vs. target architecture mismatch
  If you encounter an error missing from this list, please file an issue or a PR!', /Users/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/ir/context.rs:535:15
  stack backtrace:
     0: _rust_begin_unwind
     1: core::panicking::panic_fmt
     2: core::option::expect_failed
     3: core::option::Option<T>::expect
               at /Users/runner/work/idf-rust-build/idf-rust-build/library/core/src/option.rs:351:21
     4: bindgen::ir::context::BindgenContext::new
               at /Users/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/ir/context.rs:529:13
     5: bindgen::Bindings::generate
               at /Users/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/lib.rs:2272:27
     6: bindgen::Builder::generate
               at /Users/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.58.1/src/lib.rs:1434:9
     7: build_script_build::run_bindgen
               at ./build.rs:50:20
     8: build_script_build::main
               at ./build.rs:12:5
     9: core::ops::function::FnOnce::call_once
               at /Users/runner/work/idf-rust-build/idf-rust-build/library/core/src/ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
ninja: build stopped: subcommand failed.
ninja failed with exit code 1

alex@my-mac rust-esp32-example-main % rustc --version
rustc 1.54.0-dev

alex@my-mac rust-esp32-example-main % rustc --print target-list | grep esp
riscv32imac-esp-espidf
riscv32imc-esp-espidf
xtensa-esp32-espidf
xtensa-esp32-none-elf
xtensa-esp32s2-espidf
xtensa-esp32s2-none-elf
xtensa-esp8266-none-elf

alex@my-mac rust-esp32-example-main % clang --version
Apple clang version 12.0.5 (clang-1205.0.22.11)
Target: x86_64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

alex@my-mac rust-esp32-example-main % xtensa-esp32-elf-gcc --version
xtensa-esp32-elf-gcc (crosstool-NG esp-2020r3) 8.4.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

--- a/docs/rust-on-xtensa.md
+++ b/docs/rust-on-xtensa.md
@@ -50,6 +50,36 @@ Solution:
 
 Change the target to `xtensa-esp32-espidf`.
 
+### Build fails with panic at `libclang error`
+
+If you see the following error:
+
+```
+error: failed to run custom build command for `rustlib v0.1.0 (/home/brian/src/esp32-rust/rust-esp32-example/components/rustlib)`
+
+Caused by:
+  process didn't exit successfully: `/home/user/projects/rust-esp32-example/build/esp-idf/rustlib/target/release/build/rustlib-435225cb26b45082/build-script-build` (exit status: 101)
+  --- stdout
+  cargo:rerun-if-changed=/home/user/projects/rust-esp32-example/build/esp-idf/rustlib/target/RustApi.h
+
+  --- stderr
+  thread 'main' panicked at 'libclang error; possible causes include:
+  - Invalid flag syntax
+  - Unrecognized flags
+  - Invalid flag arguments
+  - File I/O errors
+  - Host vs. target architecture mismatch
+[...]
+```
+
+This likely means that the build script is finding the wrong `libclang.so` (or `libclang.dylib` on macOS) file (possibly from your system's global copy of clang).  To teach the build where the correct copy of libclang is, set in your environment:
+
+```
+export LIBCLANG_PATH=/path/to/xtensa-esp32-elf-clang/lib/libclang.so.12
+```
+
+(Replace the path above with the correct path on your system.)
+
 ## Building from the scratch
 
 Following text describes the build process when building LLVM and Rust from the scratch.

--- a/docs/rust-on-xtensa.md
+++ b/docs/rust-on-xtensa.md
@@ -80,9 +80,9 @@ export LIBCLANG_PATH=/path/to/xtensa-esp32-elf-clang/lib/libclang.so.12
 
 (Replace the path above with the correct path on your system.)
 
-## Building from the scratch
+## Building from scratch
 
-Following text describes the build process when building LLVM and Rust from the scratch.
+Following text describes the build process when building LLVM and Rust from scratch.
 
 ## Using Rust for ESP32 Development
 

